src/redux-enhancers/undoRedoEnhancer.ts

```diff
@@ -354,9 +388,36 @@ const undoRedoReducerEnhancer: StoreEnhancer<any> =
       if (
         (isNavigation(actionType) && isNavigation(lastActionType)) ||
         (lastActionType === 'editThought' && actionType === 'editThought') ||
-        actionType === 'closeAlert'
+        actionType === 'closeAlert' ||
+        // For multicursor operations, group all undoable actions
+        (isMulticursorExecuting && isUndoable(actionType))
       ) {
-        lastActionType = actionType
+        // Handle multicursor operations
+        if (isMulticursorExecuting) {
```

Comment:
You've created a separate branch of code to handle the multicursor, with its own return statement, even though it is doing the largely the same things as the navigation grouping: calculating a diff and then replacing the last undo patch. 

It would be better consolidate this logic, preferably after the additional state variables have been cleaned up.